firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/UpdateApkClient.java

```diff
@@ -0,0 +1,290 @@
+      throws FirebaseAppDistributionException {
+    HttpsURLConnection httpsURLConnection;
+
+    try {
+      URL url = new URL(downloadUrl);
+      httpsURLConnection = (HttpsURLConnection) url.openConnection();
+    } catch (IOException e) {
+      throw new FirebaseAppDistributionException(
+          Constants.ErrorMessages.NETWORK_ERROR, NETWORK_FAILURE, e);
+    }
+    return httpsURLConnection;
+  }
+
+  private void setDownloadTaskCompletionError(FirebaseAppDistributionException e) {
+    if (downloadTaskCompletionSource != null
+        && !downloadTaskCompletionSource.getTask().isComplete()) {
+      downloadTaskCompletionSource.setException(e);
+    }
+  }
+
+  private void setDownloadTaskCompletionErrorWithDefault(
+      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
+    if (e instanceof FirebaseAppDistributionException) {
+      setDownloadTaskCompletionError((FirebaseAppDistributionException) e);
+    } else {
+      setDownloadTaskCompletionError(defaultFirebaseException);
+    }
+  }
+
+  private Task<Void> install(String path, Activity currentActivity) {
+    Intent intent = new Intent(currentActivity, InstallActivity.class);
+    intent.putExtra("INSTALL_PATH", path);
+    CancellationTokenSource installCancellationTokenSource = new CancellationTokenSource();
+    this.installTaskCompletionSource =
+        new TaskCompletionSource<>(installCancellationTokenSource.getToken());
+    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    currentActivity.startActivity(intent);
+    return installTaskCompletionSource.getTask();
+  }
+
+  void setInstallationResult(int resultCode) {
+    if (resultCode == Activity.RESULT_OK) {
+      installTaskCompletionSource.setResult(null);
+    } else if (resultCode == Activity.RESULT_CANCELED) {
+      installTaskCompletionSource.setException(
+          new FirebaseAppDistributionException(
+              Constants.ErrorMessages.UPDATE_CANCELED,
+              FirebaseAppDistributionException.Status.INSTALLATION_CANCELED));
+    } else {
+      installTaskCompletionSource.setException(
+          new FirebaseAppDistributionException(
+              "Installation failed with result code: " + resultCode,
+              FirebaseAppDistributionException.Status.INSTALLATION_FAILURE));
+    }
+  }
+
+  private void setUpdateAppTaskCompletionError(FirebaseAppDistributionException e) {
+    if (updateAppTaskCompletionSource != null
+        && !updateAppTaskCompletionSource.getTask().isComplete()) {
+      updateAppTaskCompletionSource.setException(e);
+    }
+  }
+
+  private void setUpdateAppErrorWithDefault(
+      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
+    if (e instanceof FirebaseAppDistributionException) {
+      setUpdateAppTaskCompletionError((FirebaseAppDistributionException) e);
+    } else {
+      setUpdateAppTaskCompletionError(defaultFirebaseException);
+    }
+  }
+
```

Comment:
nit: might be worth renaming this to postProgress or postDownloadProgress

Reply:
called it postUpdateProgress since that's the new name for updateState we are gonna change